src/main/java/ir/sahab/nexus/plugin/tag/internal/TagEntityAdapter.java

```diff
@@ -0,0 +1,174 @@
+                .collect(Collectors.toList());
+        oDocument.field(COMPONENTS_FIELD, componentDocuments);
+    }
+
+    private static ODocument toDocument(AssociatedComponent component) {
+        ODocument document = new ODocument();
+        document.field(COMPONENT_REPOSITORY_FIELD, component.getRepository());
+        document.field(COMPONENT_GROUP_FIELD, component.getGroup());
+        document.field(COMPONENT_NAME_FIELD, component.getName());
+        document.field(COMPONENT_VERSION_FIELD, component.getVersion());
+        return document;
+    }
+
+    private static AssociatedComponent toComponent(ODocument document) {
+        AssociatedComponent component = new AssociatedComponent();
+        component.setRepository(document.field(COMPONENT_REPOSITORY_FIELD));
+        component.setGroup(document.field(COMPONENT_GROUP_FIELD));
+        component.setName(document.field(COMPONENT_NAME_FIELD));
+        component.setVersion(document.field(COMPONENT_VERSION_FIELD));
+        return component;
+    }
+
+    public Optional<TagEntity> findByName(ODatabaseDocumentTx tx, String name) {
+        OIndex<?> nameIndex = tx.getMetadata().getIndexManager().getIndex(NAME_INDEX);
+        OIdentifiable identifiable = (OIdentifiable) nameIndex.get(name);
+        if (identifiable == null) {
+            return Optional.empty();
+        }
+        return Optional.of(transformEntity(identifiable.getRecord()));
+    }
+
+    public Iterable<TagEntity> search(ODatabaseDocumentTx tx, Map<String, String> attributes) {
+        List<QueryPredicate> predicates = new ArrayList<>();
+        for (Entry<String, String> entry : attributes.entrySet()) {
+            String field = ATTRIBUTES_FIELD + "[" + stringLiteral(entry.getKey()) + "]";
+            predicates.add(new QueryPredicate(field, "=", stringLiteral(entry.getValue())));
+        }
+
+        String query = buildQuery(predicates);
+        log.info("Searching for tags with query={}", query);
+        List<ODocument> documents = tx.query(new OSQLSynchQuery<>(query));
+        return transform(documents);
+    }
+
+    /**
+     * @return SQL string literal for given value (enclosed in single quotes)
+     */
+    private static String stringLiteral(String value) {
+        return "'" + value + "'";
+    }
+
+    private static String buildQuery(List<QueryPredicate> predicates) {
+        StringBuilder query = new StringBuilder("select * from ").append(DB_CLASS);
+        Iterator<QueryPredicate> it = predicates.iterator();
+        if (it.hasNext()) {
+            query.append(" where");
+        }
+        while (it.hasNext()) {
+            QueryPredicate predicate = it.next();
+            query.append(' ').append(predicate.field)
+                    .append(' ')
+                    .append(predicate.operator)
+                    .append(' ')
+                    .append(predicate.value);
+            if (it.hasNext()) {
+                query.append(" and");
+            }
+        }
```

Comment:
Actually that implementation creates one string builder for query generation and that was the reason I did that way, but as it may not be important, I changed to ur suggestion.